### DISCUSSION TITLE
Multiple bars in a single plot

### DISCUSSION THREAD

**Discussion by PatrickBuTaxdoo** [observablehq/plot#1032] [Q&A] [answered]
Hello!

I want to display multiple bars in a single plot. I cannot find anything related in the docs, and also no example for that.
Adding multiple bars is not an issue by itself, but the bars are overlapping, and I cannot find an option to adjust the width of the bars. This is what I am getting:

![grafik](https://user-images.githubusercontent.com/26573439/186471134-12996ab3-4c72-45d2-b191-b2e8ab354ea3.png)

using this code:

```ts
Plot.plot({
    marks: [
      Plot.barX(data, {
        y: 'stat',
        x: 'old',
        fill: '#4e79a7',
      }),
      Plot.barX(data, {
        y: 'stat',
        x: 'new',
        height: 5,
        fill: '#cc5555',
        dy: 10,
      }),
    ],
    x: {grid: true, label: 'Time (seconds)'},
    y: {grid: true, label: 'Metric'},
  });
```

How can I plot multiple bars, or just adjust the width/height of the bars?

`strokeWidth`, `strokeHeight`, `height` and `width` on the `barX` options do not seem to work.

**Comment by yurivish** [comment]
I think you could use `insetTop` and `insetBottom` to adjust the height of the bars? Another approach would be to try [faceting](https://observablehq.com/@observablehq/plot-facets) to make mini-charts each containing two bars. I don't immediately have a reference example for that idea, but it would let you show each blue bar next to its red counterpart.

**Comment by mbostock** [comment, marked as answer]
I recommend using faceting for this, and using a “tidy” representation of data where there is one row per observation, rather than treating the “new” and “old” observations as separate dimensions. To tidy the data, you could say:

```js
tidy = data.flatMap((d) => [
  { stat: d.stat, series: "old", time: d.old },
  { stat: d.stat, series: "new", time: d.new }
])
```

Then to use faceting, you could say:

<img width="664" alt="Screen Shot 2022-08-24 at 1 18 14 PM" src="https://user-images.githubusercontent.com/230541/186515208-f223ce18-e6ef-4204-9082-92b7c32469fe.png">

```js
Plot.plot({
  x: { grid: true, label: "Time (seconds)" },
  y: { label: "Metric", axis: null },
  color: { legend: true },
  facet: {
    data: tidy,
    y: "stat",
    marginLeft: 60
  },
  marks: [
    Plot.barX(tidy, {
      x: "time",
      y: "series",
      fill: "series"
    })
  ]
})
```

Live notebook: https://observablehq.com/d/47d887315e24c39b

**Reply by PatrickBuTaxdoo** [reply]
Thank you very much!

I wanted to have the colors customizable and als add labels. I also fixed the label for the left axis. This is my final code:

```
    const tidy = data.flatMap((d) => [
      { stat: d.stat, series: 'old', time: d.old, fill: '#4f79a7' },
      { stat: d.stat, series: 'new', time: d.new, fill: '#f18e2c' },
    ]);
    Plot.plot({
      x: { grid: true, label: 'Time (seconds)' },
      y: { axis: null },
      color: { legend: true },
      facet: {
        data: tidy,
        y: 'stat',
        marginLeft: 60,
        marginRight: 60,
        label: 'Metric'
      },
      marks: [
        Plot.barX(tidy, {
          x: 'time',
          y: 'series',
          fill: 'fill',
        }),
        Plot.text(tidy, {
          y: 'series',
          x: 'time',
          text: 'time',
          textAnchor: 'start',
          dx: 10,
        }),
      ],
    });
```

**Reply by PatrickBuTaxdoo** [reply]
Unfortunately, like in your graph, the statistics are not in original order anymore ("maximum" should be last). Is there a way to keep the original order? It looks like the data is sorted alphabetically

**Reply by Fil** [reply]
you can order the facet domain by specifying it explicitly, for example:
 `fy: {domain: ["average", "p50", "p90", "p95", "maximum"]}`

**Reply by PatrickBuTaxdoo** [reply]
Awesome, thank you, that works!

**Comment by mbostock** [comment]
Another approach would be to make the “new” values more prominent (i.e., you want to see the new values in the context of the old values), and then show the change with a link or arrow.

![untitled (54)](https://user-images.githubusercontent.com/230541/186516850-7c2b32c4-9866-46fa-a6cf-c8d1041611da.png)

```js
Plot.plot({
  x: { grid: true, label: "Time (seconds)" },
  y: { label: "Metric" },
  marks: [
    Plot.barX(data, { x: "new", y: "stat", fill: "#ccc" }),
    Plot.ruleX([0]),
    Plot.tickX(data, { x: "new", y: "stat", strokeWidth: 1.5 }),
    Plot.arrow(data, { x1: "old", x2: "new", y: "stat", stroke: "red" }),
    Plot.dot(data, { x: "old", y: "stat", fill: "red" })
  ]
})
```

Shown here: https://observablehq.com/d/47d887315e24c39b

**Comment by dgwyer** [comment]
@PatrickBuTaxdoo As pointed out above you could use inset to overlay the bars too like this (needs tweaking probably): https://observablehq.com/d/20c995d04b71e4db#cell-8

I'm not sure why the legend isn't displaying though. Margin/padding needs adding?

![image](https://user-images.githubusercontent.com/1482075/186626619-331458d0-4563-4fa2-ac40-8d359c53d15d.png)

**Reply by Fil** [reply]
yes, see the marginLeft setting in Mike's notebook https://observablehq.com/d/47d887315e24c39b

**Reply by dgwyer** [reply]
Hmm, I just tried `marginLeft` (and other margins) but it didn't reveal the legend. Maybe some other setting is needed?

https://observablehq.com/d/20c995d04b71e4db#cell-8

**Reply by Fil** [reply]
can you publish the notebook?

**Reply by dgwyer** [reply]
Oops. Done. :)

**Reply by Fil** [reply]
it seems to work with:
```
Plot.plot({
  marginLeft: 80,
  x: { grid: true, label: "Time (seconds)" },
  y: { label: "Metric" },
  marks: [
    Plot.barX(data, { x: "new", y: "stat", fill: "#ccc" }),
    Plot.ruleX([0]),
    Plot.tickX(data, { x: "new", y: "stat", strokeWidth: 1.5 }),
    Plot.arrow(data, { x1: "old", x2: "new", y: "stat", stroke: "red" }),
    Plot.dot(data, { x: "old", y: "stat", fill: "red" })
  ]
})
```

**Reply by dgwyer** [reply]
The legend? I still don't see it, unless you're referring to something else. I was referring to the series legend that displays at the top of the chart.

**Reply by Fil** [reply]
Ah sorry, you meant the color legend, I thought you were speaking about the labels for y.

The marks in this chart are using fixed colors (that are not data-driven), so there is no way Plot could derive a legend in this case.

In Mike's example above, the color is data-driven (based on the "series" accessor), which allows Plot to create a _color_ scale, and a legend is then created with the _legend: true_ option.

**Reply by dgwyer** [reply]
Makes sense, thanks.
